Cls lbl prog fix v2

hecon5 opened the pull request:
Updates to clsLblProg based on additional discussions and tweaking.

joyfullservice commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -129,6 +141,28 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
+    this.blnHideCaption = NewVal
+End Property
+
+Public Property Get DisplayTimeRemaining() As Boolean
+    DisplayTimeRemaining = this.blnShowTimeRemain
+End Property
+Public Property Let DisplayTimeRemaining(NewVal As Boolean)
+    this.blnShowTimeRemain = NewVal
+    If NewVal Then this.blnShowTimeComplete = False
+End Property
+Public Property Get DisplayTimeComplete() As Boolean
+    DisplayTimeComplete = this.blnShowTimeComplete
+End Property
+Public Property Let DisplayTimeComplete(NewVal As Boolean)
+    this.blnShowTimeComplete = NewVal
+    If NewVal Then this.blnShowTimeRemain = False
+End Property
```
I'd like to continue the discussion on these, and more fully understand your perspective...

When I review the usage of progress bars in UX design, including a percentage caption was something that Microsoft did in Windows 95:

![image](https://user-images.githubusercontent.com/7254992/169667061-fa85cb24-f7de-4244-908d-381a2498def7.png)

But when I review almost any [modern professional application](https://docs.microsoft.com/en-us/windows/win32/uxguide/progress-bars), I almost never see a caption at all. Status text and percentage indications are now separate labels outside the progress bar.

![image](https://user-images.githubusercontent.com/7254992/169667193-d5143054-065a-4018-b751-e6fe8c0bdea0.png)

The latest Microsoft [guidelines](https://docs.microsoft.com/en-us/windows/apps/design/controls/progress-controls) show even smaller progress bars. Personally, I am not sure that I like these better than the Windows 7 style, but that's just my personal taste.

![image](https://user-images.githubusercontent.com/7254992/169667517-ddfedb21-79c5-4500-a974-015002cae010.png)

Again, this is my opinion here, but I feel like adding additional content into the progress bar caption is a non-standard design practice that is counter to the major UX design guidelines out there. I feel like in the interest of professionalism it would be good to stick with the tried and true designs that encourage users towards industry-standard usability principles.

That being said, developers are not limited by the progress bar class because they can simply set the caption label text to anything they want with a single line of code following the `.Increment` call.

```vba
cProg.Increment
lblCaption.Caption = "It's okay, you really only have a couple thousand more iterations to go! " & cProg.EstTimeRemaining
```

hecon5 replied:
Hmm. This opens up some interesting ideas. 

I'll cogitate on this overnight and get back on this with some final thoughts, but for now, some musings:

1. The progress class moves the caption over the bar as of now, and after reading this, I'm not sure it should. Instead, perhaps we make it the same length as the bar, but don't move it. 
2. I would prefer to address all elements of a progress bar as a single element and not individually. If you're talking to the progress bar, it should always be 
```VBA
    cprog.something
    cprog.Increment
    cprog.caption = "blah blah blah"
' (or with directive)

' this is less clean or readable IMHO

    cprog.increment
    Lblcaption.caption = "blah blah blah"
    Someothercontrol.operarion
``` 
This keeps all the time and elements together under one roof and it's clear you're working on the bar. 

3. Alternatively, we could have them on two or more lines of text like the win 7 docs suggest, and have them be multiple properties and let the class deal with it.
4. Reason being I want a self contained object I don't need to remember I have to (separately) handle other information about (this object).
5. I think the latest guidance is a tad too spartan for my tastes, too, and think while it looks clean, it gives so little context as to be no more useful than a spinning wheel .gif.
6. I do think, though, we should move the caption off the bar and put it above or below. That would also have the benefit of not needing to determine if the text should be white or black at the 50% mark...

hecon5 replied:
Ok, so after thinking more about it, I think we should get rid of the percentage text completely. 

After absorbing the UX documents for a while, the whole point of the progress bar is to show the progress. Having an additional percentage seems duplicative after consideration. 

I do think, however, we can / ?should? have a caption and the option to show the time.

Leaving the external event may allow us to enable devs to add an event to handle the caption as needed and reduce co-dependencies (say, if you wanted to show the current `perf.Category` and the current `perf.operation`, we don't want to create a dependency in the class, so handling it outside with an example may allow a clean way to do that).

hecon5 replied:
After playing around with this some more, IMO, we should not mess with the theme elements set by the developers. Size, placement, and fill color, yes, but IMO, the borders, effects, and the like should not be altered by the class. This allows devs to set the look to be consistent with the context of the control, and not need to futz with creating custom elements or internal routines for each bar in use.

joyfullservice reviewed (changes requested):
I added a few comments to the PR. Let me know what you think!  😄

joyfullservice commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -65,7 +65,7 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
 '
 #If APPLICATION_NAME = "Microsoft Access" Then
-    Public Sub Initialize(BackLabel As Label, Optional FrontLabel As Label, Optional CaptionLabel As Label)
+    Public Sub Initialize(BackLabel As Label, FrontLabel As Label, Optional CaptionLabel As Label)
```
Good catch! It was optional from when I was using the same declaration for both Access and Excel. But with the split using the compiler directive, it totally makes more sense to make this required.

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -118,6 +129,7 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
             .Width = this.lblBack.Width - 4
             .Height = this.lblBack.Height - 4
             .TextAlign = 2 'fmTextAlignCenter
+            .TopMargin = this.sngOffset * 2
```
We do, especially since the caption is placed over the bar, and in some cases the top of the text gets cut off by the sunken effect.

More on that in a minute.

hecon5 replied:
If we move the caption off the bar, this won't be needed.

joyfullservice replied:
In most cases this is probably helpful, and for the exceptions, the developer can adjust the `.TopMargin` after initializing the progress bar. I am fine with leaving this in with the compiler directive.

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -129,6 +141,28 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
 
 End Sub
 
+Public Property Get HideCaption() As Boolean
+    HideCaption = this.blnHideCaption
+End Property
+Public Property Let HideCaption(NewVal As Boolean)
+    this.blnHideCaption = NewVal
+End Property
```
We actually do use it fairly often, for instance we show the perf.category or other large operations, but no text on a cleanup to reduce interruption time.

hecon5 replied:
Revised to new name `CaptionVisible`. That was done to ensure consistent logic and naming with other areas.

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -20,6 +20,8 @@ Option Explicit
 ' object types in early binding.
 #Const APPLICATION_NAME = "Microsoft Access"
 
+Private Const lngBarColorDefault As Long = 8388608 ' Default progress bar color (Navy Blue). 
```
I'm wondering if we should make the color a `property` instead. Would allow dynamic color changes, and using either the external event you included or an internal timer set to 50% more than the average increment may allow one to detect stalled progress and change the color of the bar. (Think file copy stalled progress or conflicts)

hecon5 replied:
This also reduces the number of arguments we "must" provide to initiate, which IMHO is a cleaner thing, even if there's the additional code of the property.

joyfullservice replied:
Yes, I like the idea of a property for this. 👍

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -105,22 +118,21 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
-            .BackStyle = 0 'fmBackStyleTransparent
-            .Caption = "0%"
-            .ForeColor = 16777215   ' white
+            .Height = (this.lblBack.Height - 4) * 2 ' Two lines of text should be shown
```
I'm not sure this is correct, but initial testing shows it is close

hecon5 replied:
Per other discussions, this isn't needed since the Caption label isn't on the bar any longer.

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -105,22 +118,21 @@ Public Event ProgressChange(dblValue As Double, intPercent As Integer)
+            .Height = (this.lblBack.Height - 4) * 2 ' Two lines of text should be shown
+            .TextAlign = 1 ' fmTextAlignLeft
+            .BackStyle = 0 ' fmBackStyleTransparent
+            .Caption = vbNullString
```
With the new method of caption display and UX discussion, IMHO we shouldn't show the percent anymore.

hecon5 replied:
![image](https://user-images.githubusercontent.com/54177882/169848179-ac56ca74-c6cc-4cfa-bb8c-f5be99280d4a.png)
Here's a rendering option for the display bar.

hecon5 replied:
I think I need to tweak the text size and the like, but it looks a LOT better than being on the bar, even if it does take a bit more room.

hecon5 replied:
Toying around with some effects: flat with a slight hairline border looks ok.
![image](https://user-images.githubusercontent.com/54177882/169925815-dac1d96f-8066-4045-b2de-d12f21098bf0.png)

Increasing the size of the text to 10 pt.
![image](https://user-images.githubusercontent.com/54177882/169926006-26d7c1e2-43ab-4621-b646-359aa9050307.png)

hecon5 replied:
And pretty close to Window's default green: 
![image](https://user-images.githubusercontent.com/54177882/169926159-ea14e99e-87dc-43a7-bc84-98c792399ccb.png)

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -359,19 +433,11 @@ Private Sub UpdateDisplay(Optional blnForceUpdate As Boolean)
+	    End If
     End With
 
-    ' Use white or black, depending on progress level
-    If Me.Value > (Me.Max / 2) Then
-        ' Show white on blue background after 50%
-        this.lblCaption.ForeColor = 16777215
-    Else
-        ' Show black on grey background
-        this.lblCaption.ForeColor = 0
-    End If
-
```
We shouldn't have the percent on the bar, therefore this is not needed anymore.

hecon5 commented:
Ok, so, I think this is pretty close. I fixed some white space issues and also updated the hiding function to avoid needing to hide errors (by not actually creating them in the first place...)

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -400,8 +468,11 @@ Private Function GetParentFormName(ctlControl As Control) As String
 
     Set objParent = ctlControl
-
+#If APPLICATION_NAME = "Microsoft Access" Then
+    Do While Not TypeOf objParent Is Form
+#Else
     Do While Not TypeOf objParent Is UserForm
+#End If
```
IF that's the case, we could also use a compiler directive to ignore the entire function,

hecon5 replied:
Pushed a potential fix.

hecon5 commented:
Found some edge cases where if you load a Max Value of less than the current value it'll overflow the bar. 

I'm wondering which of these make more sense: 

1. Set value to Max if the value is greater than new max value. I don't like this much as it doesn't allow for OTAAT.
Looks like this:
```VBA
Public Property Let Max(NewVal As Double)
    this.dblMax = NewVal
    ' Prevent overflow:
    If this.dblValue > NewVal Then this.dblValue = NewVal
    UpdateDisplay True
End Property
```
3. Simply display 100% if value is greater. I prefer this, but it (may) have other implications I'm not aware of.
```VBA
' Done in "UpdateDisplay" function:
        If .dblMax <= 0 Then
            intPercent = 0
            dblWidth = 0
        ElseIf .dblValue >= .dblMax Then
            intPercent = 100
            dblWidth = .dblFullWidth
        Else
            intPercent = .dblValue * (100 / Me.Max)
            dblWidth = .dblValue * (.dblFullWidth / Me.Max)
        End If

```

joyfullservice commented:
> @joyfullservice, not sure if you had thoughts on the max value issue

Right now we have two sets of logic involved. We have the `Value` and `Max` properties (LET) and the `UpdateDisplay` function. Here is how I envision the relationship between these:

**Value/Max Properties**
This is where the current value of the progress bar is changed. The progress bar displays a ratio of a value to a whole. When either of these two properties change, this triggers a review on whether we need to update the displayed value. In my perspective, this is where we need to enforce the mathematical integrity of these values. (Value cannot exceed Max.) In a sense, we are sanitizing the incoming data so we don't encounter logic errors further down the road.

**Update Display**
This is where we evaluate the values and determine whether to update the properties of the progress bar display controls, and fire the `DoEvents` that actually repaints the progress bar on the screen. This final step should only be run when we have an actual change to show. Otherwise it slows down the processing through unneeded calls to `DoEvents`.

To fully implement what I am describing here will require some minor adjustment/refactoring of the existing code, but I feel it would organize the logic more neatly into the functions created for those purposes.

joyfullservice commented:
> @joyfullservice, not sure if you had thoughts on the max value issue

Right now we have two sets of logic involved. We have the `Value` and `Max` properties (LET) and the `UpdateDisplay` function. Here is how I envision the relationship between these:

**Value/Max Properties**
This is where the current value of the progress bar is changed. The progress bar displays a ratio of a value to a whole. When either of these two properties change, this triggers a review on whether we need to update the displayed value. In my perspective, this is where we need to enforce the mathematical integrity of these values. (Value cannot exceed Max.) In a sense, we are sanitizing the incoming data so we don't encounter logic errors further down the road.

**Update Display**
This is where we evaluate the values and determine whether to update the properties of the progress bar display controls, and fire the `DoEvents` that actually repaints the progress bar on the screen. This final step should only be run when we have an actual change to show. Otherwise it slows down the processing through unneeded calls to `DoEvents`.

To fully implement what I am describing here will require some minor adjustment/refactoring of the existing code, but I feel it would organize the logic more neatly into the functions created for those purposes.

hecon5 commented:
I like it, since `Value` already handles if it's set too high, it would make sense to manage `max` the same way. I just wanted to confirm `value` should be lowered to max in the cases of `max` being set.

This

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -337,17 +410,21 @@ End Sub
         If .dblMax <= 0 Then
             intPercent = 0
             dblWidth = 0
+        ElseIf .dblValue >= .dblMax Then
+            intPercent = 100
+            dblWidth = .dblFullWidth
```
I'm still of the opinion this should remain; after running this thing through 1000's of iterations in testing, the additional processing time is negligible and prevents overflows (if we remove the `UpdateDisplay` call after `Let Max` we're far ahead because we're cutting out potentially 3 `Do Events` if you call `Reset`). The end result is a bar which shows 100%, and that it doesn't alter values, just the display. 99.99% of the time, the calling routine will be after `Let Value` (since it's public it can't be 100%). 

Since MOST of the time `Let Max` will be called at the end of a cycle of operations (but not all for a first/second pass check), I strongly feel `Let Max` should do nothing other than update `Max` and nothing else. These lines protect from overflow, don't cost anything, and `Let Value` is the only thing which should alter `Value`.

joyfullservice replied:
That's fine with me.

hecon5 commented on Version Control.accda.src/modules/clsLblProg.cls:
```diff
@@ -171,17 +244,6 @@ Public Property Let Value(ByVal dblValue As Double)
 
     Dim dblChange As Double
 
-    ' Check visibility
-    If Me.Max = 0 Or dblValue = 0 Then
-        SetVisibility this.lblFront, False
-        SetVisibility this.lblCaption, False
-    Else
-        ' Ensure controls are visible
-        SetVisibility this.lblBack, True
-        SetVisibility this.lblFront, True
-        SetVisibility this.lblCaption, True
-    End If
-
```
Since we're handling these in `UpdateDisplay` again in a moment, there's no reason to do this here.

joyfullservice commented:
Okay, I think we probably should wrap this up so we can move on to other things...  😄  I think there are mainly two outstanding issues before I merge this in...

### Caption Placement
The goal and purpose of this class is to represent a _progress bar_, a visual indicator of the progressive completion of a multi-step process. It's not really intended to manage anything outside the scope of the actual bar. For this reason, I am good with a (simple) percent indicator as is commonly seen with progress bars, overlaid and centered on the progress bar.

I totally understand that some people are going to want the caption off, or to use a caption or status updates in another label or text box somewhere else on the front end. The idea here is that if you want the percentage shown in the traditional way (as we have by default if you pass in a label when initializing the progress bar) it simply works right out of the box. 

If you don't want the caption because you want a thin, flat progress bar, or want to display the caption somewhere else, that is great, but that is up to the front end designer. The class provides an update event to fire UI updates automatically, or the developer can simply add those next to the incrementing code in their loop. I am good with exposing the properties of estimated time remaining and estimated time to completion, but _how those values are displayed_ is up to the designer, not this class.

### Style and Colors
Here we have a balance between simplicity and flexibility. On the simplicity side, this class has historically been super easy to use. Simply drop three labels on your form, and link them to the class on the `Initialize` function. All the sizing is done based on the back label and you don't have to mess with any properties for the labels. It simply works, right out of the box. For most users, the sunken back with the blue bar serves the purpose of displaying progress and the developer can focus their time on the other elements of the application. To have to remember to set properties every time you want to add a progress bar breaks the simplicity of it.

On the other hand, I understand that people also like flexibility and options. Maybe they don't want the traditional progress bar look. In the vast majority of my actual use cases the traditional approach was just fine, but in the splash screen I use in all my applications, I wanted a more modern look with a thin white bar expanding across a flat grey bar.

![image](https://user-images.githubusercontent.com/7254992/170843677-50c1df25-1127-4a42-9cda-5cbde4120166.png)

For something like this, I don't mind exposing some properties that can be used to adjust the default appearance from the class, but I don't want to over complicate this either. We could have a style enum with various options to use, but then we start heading down the road of how many different styles do we want to include? In my opinion, this is better served by a factory/constructor function in the end-user application that initiates the class and sets those properties as desired.

```vba
Public Function NewProg(BackLabel As Label, FrontLabel As Label, Optional CaptionLabel As Label) As clsLblProg
    Set NewProg = New clsLblProg
    With NewProg
        .Initialize BackLabel, FrontLabel, CaptionLabel
        .HideCaption = True
        .BackStyle = 1
        .BarColor = 123456
        .BackColor = 987654
    End With
End Function
```

Then instead of `set cProg = New clsLblProg` you would use this: `set cProg = NewProg(lblBack, lblFront)` and you are off and running with a nice application specific styling without having to set properties or write additional code when implementing the progress bar throughout your application.

### Final Thoughts
I know you have put a lot of time and effort into these updates, so please don't take this as ungratefulness for what you have done in a few of the aspects where we came to different conclusions. As the project owner, sometimes I just have to make the tough calls to keep the project heading in the direction that I believe will best serve our users in the long run.

Feel free to make some additional changes if you like, then next week I am planning to go ahead and merge the PR and probably make some adjustments as described here. I could have done that today, but I thought it might be better to communicate a little of the why behind it before making the changes.  😄 Have a great weekend!!

hecon5 commented:
Was away for a long weekend, some r&r felt nice away from the keyboard!

> Simply drop three labels on your form, and link them to the class on the Initialize function. All the sizing is done based on the back label and you don't have to mess with any properties for the labels. It simply works, right out of the box.

I really like this ease; I struggled with how best to handle a caption component as well, because if you have a modern bar, the caption can't live on it; and if you have a caption, it would grow potentially onto some other control or element, so having it not auto-size was the compromise.

>  I wanted a more modern look with a thin white bar expanding across a flat grey bar.

I think this is what I'm after for most of ours, now, and is why I separated out the caption. Having said that, I can see where you're going with the caption on the bar. Given most progress bars are now "modern", I think it gets a bit more difficult. 

>  In my opinion, this is better served by a factory/constructor function in the end-user application that initiates the class and sets those properties as desired.

I think that's a good idea; balances ease of building and utility.

> so please don't take this as ungratefulness for what you have done in a few of the aspects where we came to different conclusions 

Definitely don't. I'm fully aware my "ideas" aren't always going to be used. That's part of life and dialogue.

> Have a great weekend!!

I for sure did.

@joyfullservice, I think I'm done making changes with the addition of the `BackColor` property (for `lblBack`). While I could tweak this forever, I think with where it is I'm happy with it. I think adding a factory constructor would be a good change, but since that could ostensibly be built outside the class, I think it's not required. In all, if I feel we need some feature for my team's tools, I can always just fork it anyway.

hecon5 commented:
I will say, that splash screen is really nice, and very clean; assuming it's Access, it's VERY clean! I like it!
